### PR TITLE
Add Tween.reverse and some features for Tween

### PR DESCRIPTION
**Pull request by dumganhar**
Discussion: https://forum.cocos.org/t/topic/157814

Re: #17035

- Add Tween.reverse to reverse an action in the current tween or a specific tween
- Add Tween.id(number) to mark the previous action
- Add the parameter for Tween.union(fromId: number) to union the internal actions from the action with the ID
- Doesn't need to pass a target parameter for Tween.clone() if the same target is required

### Tween.id

```ts
const t = tween(node)
    .by(1, { position: new Vec3(100, 0, 0) }).id(123) // Mark the previous `by` action with 123
    .to(1, { position: new Vec3(1000, 0, 0) });
```

### Tween.reverse

1. Reverse the whole tween
```ts
const t = tween(node).by(1, { position: new Vec3(100, 0, 0) });

tween(node)
    .to(1, { position: new Vec3(200, 0, 0) })
    .reverse(t) // Reverse all actions in t
    .start();
```

2. Reverse an action by id in another tween

```ts
const node = new Node();

const t = tween(node)
    .by(1, { position: new Vec3(100, 0, 0) }).id(123)
    .to(1, { position: new Vec3(1000, 0, 0) });

tween(node)
    .to(1, { position: new Vec3(200, 0, 0) })
    .reverse(t, 123) // Only reverse `.by(1, { position: new Vec3(100, 0, 0) })` in t
    .start();
```

3. Reverse an action by id in current tween

```ts
const node = new Node();

tween(node)
    .to(1, { scale: new Vec3(10, 10, 10) })
    .by(1, { position: new Vec3(200, 0, 0) }).id(123)
    .delay(1)
    .reverse(123) // Revert the `by` action in the current tween.
    .start();
```

### Tween.union(fromId?: id)

```ts
    const node = new Node();

    tween(node)
        .to(1, { scale: new Vec3(10, 10, 10) })
        .by(1, { position: new Vec3(200, 0, 0) }).id(123)
        .delay(1)
        .reverse(123)
        .union(123) // Union from `.by(1, { position: new Vec3(200, 0, 0) })`, ignore the first `.to(...)`
        .repeat(3)
        .start();
```

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/tween/tween-action.ts [hunk 1]
```diff
@@ -161,9 +162,29 @@ export class TweenAction<T> extends ActionInterval {
         this.initWithDuration(duration);
     }
 
+    get relative (): boolean {
+        return !!this._opts!.relative;
```
Comment (minggo) [inline cocos/tween/tween-action.ts:166]: Why use !!?

File: cocos/tween/tween-action.ts [hunk 2]
```diff
@@ -161,9 +162,29 @@ export class TweenAction<T> extends ActionInterval {
...
+        action.workerTarget = this.workerTarget;
+        action._id = this._id;
+        this._cloneDecoration(action);
+        return action;
+    }
+
+    reverse (): TweenAction<T> {
+        if (!this._opts!.relative) {
+            warn('reverse: could not reverse a non-relative action');
+            return this.clone();
```
Comment (minggo) [inline cocos/tween/tween-action.ts]: Is it ok to add a dummy tween action.

File: cocos/tween/tween-action.ts (hunk 1 above)
Comment (dumganhar) [reply cocos/tween/tween-action.ts:166]: Since `relative` maybe `undefined`.

File: cocos/tween/tween-action.ts (hunk 2 above)
Comment (dumganhar) [reply cocos/tween/tween-action.ts]: Good point. I think we could new an empty TweenAction here.

File: cocos/tween/tween.ts [hunk 3]
```diff
@@ -93,6 +107,98 @@ export class Tween<T extends object = any> {
...
+     * @note @en If `id` is not assigned or its value is -1, it will reverse all actions in the `otherTween`
+     *       @zh 如果 `id` 没有被指定或者值为 -1，那么将翻转 `otherTween` 中所有的动作
+     */
+    reverse<U extends object = any> (otherTween: Tween<U>, id?: number): Tween<T>;
+
+    reverse<U extends object = any> (otherTweenOrId?: Tween<U> | number, id?: number): Tween<T> {
+        if (otherTweenOrId == null && id == null) {
+            if (this._actions.length === 0) {
+                warn('reverse: current tween could not be reversed, empty actions');
+                return this.clone(this._target as T);
```
Comment (minggo) [inline cocos/tween/tween.ts]: Is it better to dummy tween action as the comment says.

File: cocos/tween/tween.ts [hunk 4]
```diff
@@ -93,6 +107,98 @@ export class Tween<T extends object = any> {
...
+     * @param otherTween @en The tween in which to find the action by ID
+     *                   @zh 根据标识在关联的缓动中查找动作
+     * @param id @en The ID of the action to reverse @zh 要翻转的动画标识
+     * @return @en The current tween instance @zh 当前缓动实例
+     * @note @en If `id` is not assigned or its value is -1, it will reverse all actions in the `otherTween`
+     *       @zh 如果 `id` 没有被指定或者值为 -1，那么将翻转 `otherTween` 中所有的动作
+     */
+    reverse<U extends object = any> (otherTween: Tween<U>, id?: number): Tween<T>;
+
+    reverse<U extends object = any> (otherTweenOrId?: Tween<U> | number, id?: number): Tween<T> {
```
Comment (minggo) [inline cocos/tween/tween.ts:135]: I think -1 is an invalid value. So should not support it.

File: cocos/tween/tween.ts [hunk 5]
```diff
@@ -93,6 +107,98 @@ export class Tween<T extends object = any> {
...
+        if (otherTweenOrId instanceof Tween) {
+            otherTween = otherTweenOrId;
+            if (id !== undefined) {
+                actionId = id;
+            }
+        } else if (typeof otherTweenOrId === 'number') {
+            actionId = otherTweenOrId;
+        }
+
+        const actions = otherTween ? otherTween._actions : this._actions;
```
Comment (minggo) [inline cocos/tween/tween.ts]: If otherTween is number, then it may have problem.

File: cocos/tween/tween.ts (hunk 3 above)
Comment (dumganhar) [reply cocos/tween/tween.ts]: If `this._action.length` is null, Tween.clone will return an empty Tween.

File: cocos/tween/tween.ts [hunk 6]
```diff
@@ -93,6 +107,98 @@ export class Tween<T extends object = any> {
...
+        let action: FiniteTimeAction | null = null;
+        if (actions.length === 0) {
+            return this;
+        }
```
Comment (minggo) [inline cocos/tween/tween.ts:162]: Put the codes follow `const actions = otherTween ? otherTween._actions : this._actions;` is better.

File: cocos/tween/tween.ts (hunk 4 above)
Comment (dumganhar) [reply cocos/tween/tween.ts:135]: Okay, I will add an unit test for invalid id.

File: cocos/tween/tween.ts (hunk 5 above)
Comment (dumganhar) [reply cocos/tween/tween.ts]: otherTween is impossibility a number.

Comment (minggo) [inline cocos/tween/tween.ts]: It is better to define a targetTween first, then can use it in the following codes, for example
```typescript
const targetTween = otherTweenOrId instanceof Tween ? otherTweenOrId : this;
```

File: cocos/tween/tween.ts [hunk 7]
```diff
@@ -176,22 +282,50 @@ export class Tween<T extends object = any> {
...
+        };
+
+        if (fromId === undefined) {
+            unionAll();
+            return this;
+        }
+
+        const actions = this._actions;
+        const index = actions.findIndex((action) => action.getId() === fromId);
+        if (index === -1) {
```
Comment (minggo) [inline cocos/tween/tween.ts]: I think -1 is an invalid value, don't have to support it.

File: cocos/tween/tween-action.ts (hunk 2 above)
Comment (dumganhar) [reply cocos/tween/tween-action.ts]: Updated

File: cocos/tween/tween.ts (hunk 4 above)
Comment (dumganhar) [reply cocos/tween/tween.ts:135]: Updated

File: cocos/tween/tween.ts (hunk 6 above)
Comment (dumganhar) [reply cocos/tween/tween.ts:162]: Updated.

File: cocos/tween/tween.ts (hunk 5 above)
Comment (dumganhar) [reply cocos/tween/tween.ts]: It has already been updated. I refactored a little bit.

File: cocos/tween/tween.ts (hunk 7 above)
Comment (dumganhar) [reply cocos/tween/tween.ts]: Yes, add a warn log now.
```ts
const index = actions.findIndex((action) => action.getId() === fromId);
        if (index === -1) {
            warn(`union: found: ${invalidActionId}`);
            return this;
        } else {
```

File: cocos/tween/actions/action-interval.ts [hunk 8]
```diff
@@ -407,6 +408,23 @@ export class Sequence extends ActionInterval {
             actionOne.workerTarget = workerTarget;
         }
     }
+
+    findAction (id: number): FiniteTimeAction | null {
```
Comment (minggo) [inline cocos/tween/actions/action-interval.ts:412]: I think it is better to check id first.

File: cocos/tween/actions/action-interval.ts [hunk 9]
```diff
@@ -814,6 +835,30 @@ export class Spawn extends ActionInterval {
             one.workerTarget = workerTarget;
         }
     }
+
+    findAction (id: number): FiniteTimeAction | null {
```
Comment (minggo) [inline cocos/tween/actions/action-interval.ts:839]: I think it is better to check id first.

File: cocos/tween/actions/action.ts [hunk 10]
```diff
@@ -194,6 +196,14 @@ export abstract class Action {
         this.tag = tag;
     }
 
+    setId (id: number): void {
```
Comment (minggo) [inline cocos/tween/actions/action.ts:206]: It is better to check id first.

File: cocos/tween/actions/action.ts [hunk 11]
```diff
@@ -95,6 +95,8 @@ export abstract class Action {
 
     protected tag = Action.TAG_INVALID;
 
+    protected _id = -1;
```
Comment (minggo) [inline cocos/tween/actions/action.ts]: It is better to add comment about the valid range of id.

File: cocos/tween/tween.ts [hunk 12]
```diff
@@ -79,6 +80,20 @@ export class Tween<T extends object = any> {
         return this;
     }
 
+    /**
+     * @en Set the id for previous action
+     * @zh 设置前一个动作的 id
+     * @param id @en The internal action id to set @zh 内部动作的 id 标识，
+     * @note @en Don't pass in a value -1 if you want to mark the previous action since -1 is the default value
+     *       @zh 如果你想标识前一个动作，请勿传递 -1，因为 -1 是所有动作的默认标识
```
Comment (minggo) [inline cocos/tween/tween.ts]: I think -1 is a invalid value, don't have to support it here.

File: cocos/tween/tween.ts [hunk 13]
```diff
@@ -93,6 +108,123 @@ export class Tween<T extends object = any> {
...
+        if (action) {
+            reversedAction = action.reverse();
+            reversedAction.workerTarget = t._target;
+        } else {
+            warn(`reverse: could not find action id ${actionId}`);
+        }
+        return reversedAction;
+    }
+
+    private findAction (id: number, actions: FiniteTimeAction[]): FiniteTimeAction | null {
```
Comment (minggo) [inline cocos/tween/tween.ts:199]: I think should check id first.

Comment (dumganhar) [reply cocos/tween/tween.ts:199]: Fixed

File: cocos/tween/tween.ts (hunk 12 above)
Comment (dumganhar) [reply cocos/tween/tween.ts]: Fixed.

File: cocos/tween/actions/action.ts (hunk 11 above)
Comment (dumganhar) [reply cocos/tween/actions/action.ts]: Fixed.

File: cocos/tween/actions/action.ts (hunk 10 above)
Comment (dumganhar) [reply cocos/tween/actions/action.ts:206]: Currently, -1  could not be passed to Action instance by `setId` now, Tween.union, Tween.reverse, Tween.id will return failure if passing -1 to them. Therefore, I don't think we need to check whether id is -1 in `Action.setId` method. More, setId(-1) could be used to reset the id to the default value -1.

Comment (minggo) [reply cocos/tween/actions/action.ts:206]: But the game code may also invoke this function too.

Comment (dumganhar) [reply cocos/tween/actions/action.ts:206]: Updated, id 's default value is undefined now. And all numbers are treated as valid values.

**Review by minggo — APPROVED** [review]

**Pull request merged**